File: python/heritage/domain/risk_policy.py
# ============================================================
# 文件：risk_policy.py
# 职责：定义风险策略接口（IRiskPolicy）及默认实现（SimpleRiskPolicy）。
#
# 模式：策略模式（Strategy Pattern）
#   IRiskPolicy 抽象"如何根据风险指标做出决策"，
#   SimpleRiskPolicy 实现基本限额管控规则（来自 Risk_Calibration.md §1.8）。
# ============================================================

from __future__ import annotations
from abc import ABC, abstractmethod
import math

from events.events import RiskControlEvent, RiskAction
from .risk_metrics import RiskMetrics


class IRiskPolicy(ABC):
    """风险策略接口。

    输入：当前账户 ID + RiskMetrics 快照
    输出：需要执行的风控事件列表（可为空）
    """

    @abstractmethod
    def evaluate(
        self, account_id: str, metrics: RiskMetrics
    ) -> list[RiskControlEvent]:
        """评估风险指标，返回需要触发的风控事件列表。"""


class SimpleRiskPolicy(IRiskPolicy):
    """简单限额风险策略（MVP 实现）。

    规则（来自 Risk_Calibration.md §1.8）：
      1. realized_pnl < -loss_limit    → BlockOrders（止损冻结）
      2. |delta| > delta_limit         → ReduceOnly（Delta 超限）
      3. intraday_drawdown > drawdown_limit → 日志预警（不发出风控指令）
    """

    def __init__(
        self,
        loss_limit:     float = 1e6,
        delta_limit:    float = 10000.0,
        drawdown_limit: float = 5e5,
    ) -> None:
        self._loss_limit     = loss_limit      # 触发 BlockOrders 的亏损阈值
        self._delta_limit    = delta_limit     # 触发 ReduceOnly 的 Delta 绝对值上限
        self._drawdown_limit = drawdown_limit  # 触发风险预警的日内回撤上限

    def evaluate(
        self, account_id: str, metrics: RiskMetrics
    ) -> list[RiskControlEvent]:
        actions: list[RiskControlEvent] = []

        # ── 规则 1：止损触发 ─────────────────────────────
        if metrics.realized_pnl < -self._loss_limit:
            actions.append(RiskControlEvent(
                account_id = account_id,
                action     = RiskAction.BlockOrders,
                reason     = (
                    f"已实现亏损 ${int(metrics.realized_pnl)} "
                    f"超出止损限额 ${int(self._loss_limit)}"
                ),
            ))
            print(
                f"[风控策略|{account_id}] ⛔ 触发止损！"
                f" 已实现盈亏: ${metrics.realized_pnl:.2f}"
                f"  限额: -${self._loss_limit:.0f}  → BlockOrders"
            )

        # ── 规则 2：Delta 超限 ───────────────────────────
        if abs(metrics.delta) > self._delta_limit:
            actions.append(RiskControlEvent(
                account_id = account_id,
                action     = RiskAction.ReduceOnly,
                reason     = (
                    f"组合 Delta {metrics.delta:.1f} "
                    f"超出限额 ±{int(self._delta_limit)}"
                ),
            ))
            print(
                f"[风控策略|{account_id}] ⚠ Delta 超限！"
                f" Delta: {metrics.delta:.1f}"
                f"  限额: ±{self._delta_limit:.0f}  → ReduceOnly"
            )

        # ── 规则 3：日内回撤预警 ─────────────────────────
        if metrics.intraday_drawdown > self._drawdown_limit:
            print(
                f"[风控预警|{account_id}] 📉 日内回撤预警！"
                f" 回撤: ${metrics.intraday_drawdown:.2f}"
                f"  阈值: ${self._drawdown_limit:.0f}"
            )

        return actions

File: src/domain/RiskPolicy.cpp
#include "RiskPolicy.hpp"
#include <cmath>
#include <iostream>
#include <iomanip>

// ============================================================
// SimpleRiskPolicy 实现
//
// 每次 evaluate() 调用：
//   1. 检查已实现亏损是否超出止损阈值 → 发出 BlockOrders
//   2. 检查组合 Delta 是否超出方向性风险限额 → 发出 ReduceOnly
//   3. 检查日内回撤是否超出预警阈值 → 记录日志（此处通过 RiskAlertEvent）
// ============================================================

namespace omm::domain {

SimpleRiskPolicy::SimpleRiskPolicy(
    double loss_limit,
    double delta_limit,
    double drawdown_limit)
    : loss_limit_(loss_limit)
    , delta_limit_(delta_limit)
    , drawdown_limit_(drawdown_limit) {}

std::vector<events::RiskControlEvent> SimpleRiskPolicy::evaluate(
    const std::string& account_id,
    const RiskMetrics& metrics) const {

    std::vector<events::RiskControlEvent> actions;

    // ── 规则 1：止损触发 ──────────────────────────────────
    // 当已实现亏损超出阈值时，冻结账户下单权限
    if (metrics.realized_pnl < -loss_limit_) {
        actions.push_back(events::RiskControlEvent{
            account_id,
            events::RiskAction::BlockOrders,
            "已实现亏损 $" + std::to_string(static_cast<int>(metrics.realized_pnl))
            + " 超出止损限额 $" + std::to_string(static_cast<int>(loss_limit_))
        });
        std::cout << "[风控策略|" << account_id << "] ⛔ 触发止损！"
                  << " 已实现盈亏: $" << std::fixed << std::setprecision(2)
                  << metrics.realized_pnl
                  << "  限额: -$" << loss_limit_ << "  → BlockOrders\n";
    }

    // ── 规则 2：Delta 超限 ────────────────────────────────
    // 当组合方向性风险超出限额时，限制为仅允许减仓方向的订单
    if (std::abs(metrics.delta) > delta_limit_) {
        actions.push_back(events::RiskControlEvent{
            account_id,
            events::RiskAction::ReduceOnly,
            "组合 Delta " + std::to_string(metrics.delta)
            + " 超出限额 ±" + std::to_string(static_cast<int>(delta_limit_))
        });
        std::cout << "[风控策略|" << account_id << "] ⚠ Delta 超限！"
                  << " Delta: " << std::fixed << std::setprecision(1)
                  << metrics.delta
                  << "  限额: ±" << delta_limit_ << "  → ReduceOnly\n";
    }

    // ── 规则 3：日内回撤预警 ──────────────────────────────
    // 当日内回撤超出预警阈值时，记录预警（不强制动作，仅通知）
    if (metrics.intraday_drawdown > drawdown_limit_) {
        std::cout << "[风控预警|" << account_id << "] 📉 日内回撤预警！"
                  << " 回撤: $" << std::fixed << std::setprecision(2)
                  << metrics.intraday_drawdown
                  << "  阈值: $" << drawdown_limit_ << "\n";
        // 注意：此处仅打印日志，RiskAlertEvent 由 RealtimeRiskApp 独立发布
    }

    return actions;
}

} // namespace omm::domain
